File: go/tests/linkpreview_test.go
package relaytest

import (
	"net/http"
	"testing"
	"time"

	"github.com/drim-dev/verifying-agent-code/go/src/relay/app"
	"github.com/drim-dev/verifying-agent-code/go/src/relay/domain"
)

func TestS_LP_01_unfurl_success_sets_title(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	fixture.Unfurl.ProgramOK("Example")

	var msg struct {
		LinkPreviewTitle *string `json:"linkPreviewTitle"`
	}
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "see https://example.com here"}).
		expect(http.StatusCreated).decode(&msg)
	if msg.LinkPreviewTitle == nil || *msg.LinkPreviewTitle != "Example" {
		t.Fatalf("expected linkPreviewTitle Example, got %v", msg.LinkPreviewTitle)
	}
	if c := fixture.Unfurl.RequestCount(); c != 1 {
		t.Fatalf("expected exactly one unfurl request, got %d", c)
	}
}

// ---- G-HTTP: a slow/failing unfurl must degrade, never escape or hang ----

func TestS_LP_02_slow_unfurl_degrades_within_deadline(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	assertSlowUnfurlDegrades(t, fixture.BaseURL(), ada, ch)
}

func TestS_LP_02_naive_no_timeout_is_caught(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	naive := fixture.NaiveApp(func(d *app.Deps) { d.LinkPreviewer = naiveLinkPreviewer{baseURL: fixture.Unfurl.BaseURL()} })
	defer naive.Close()
	expectCatchToFail(t, "G-HTTP", func(tb testing.TB) {
		assertSlowUnfurlDegrades(tb, naive.BaseURL(), ada, ch)
	})
}

func assertSlowUnfurlDegrades(tb testing.TB, baseURL string, sender domain.User, channelID string) {
	tb.Helper()
	fixture.Unfurl.ProgramDelay(2 * time.Second)
	start := time.Now()
	resp := clientAt(tb, baseURL, sender.ID).post("/channels/"+channelID+"/messages", map[string]any{"text": "slow https://example.com"})
	elapsed := time.Since(start)
	resp.expect(http.StatusCreated)
	if elapsed > 1500*time.Millisecond {
		tb.Fatalf("post took %v, exceeding the 1.5 s degradation budget (unfurl was not bounded)", elapsed)
	}
	var msg struct {
		LinkPreviewTitle *string `json:"linkPreviewTitle"`
	}
	resp.decode(&msg)
	if msg.LinkPreviewTitle != nil {
		tb.Fatalf("a timed-out unfurl must degrade to null title, got %v", *msg.LinkPreviewTitle)
	}
}

func TestS_LP_03_server_error_degrades(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	fixture.Unfurl.ProgramServerError()
	var msg struct {
		LinkPreviewTitle *string `json:"linkPreviewTitle"`
	}
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "broken https://example.com"}).
		expect(http.StatusCreated).decode(&msg)
	if msg.LinkPreviewTitle != nil {
		t.Fatalf("a 500 unfurl must degrade to null title, got %v", *msg.LinkPreviewTitle)
	}
}

func TestS_LP_04_circuit_breaker_opens(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	ch := seedChannel(t, ada, "team", false)
	fixture.Unfurl.ProgramServerError()
	for i := 0; i < 5; i++ {
		newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "fail https://example.com"}).
			expect(http.StatusCreated)
	}
	var msg struct {
		LinkPreviewTitle *string `json:"linkPreviewTitle"`
	}
	newClient(t, ada.ID).post("/channels/"+ch+"/messages", map[string]any{"text": "sixth https://example.com"}).
		expect(http.StatusCreated).decode(&msg)
	if msg.LinkPreviewTitle != nil {
		t.Fatalf("breaker-open post should have a null title, got %v", *msg.LinkPreviewTitle)
	}
	if c := fixture.Unfurl.RequestCount(); c != 5 {
		t.Fatalf("breaker should skip the 6th outbound call: expected 5 requests, got %d", c)
	}
}

func TestS_LP_05_direct_proxy(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	fixture.Unfurl.ProgramOK("Proxied")
	var body struct {
		Title string `json:"title"`
	}
	newClient(t, ada.ID).get("/links/preview?url=https://example.com").expect(http.StatusOK).decode(&body)
	if body.Title != "Proxied" {
		t.Fatalf("expected proxied title, got %q", body.Title)
	}
	fixture.Unfurl.ProgramServerError()
	newClient(t, ada.ID).get("/links/preview?url=https://example.com").
		expect(http.StatusBadGateway).expectCode("unfurl:upstream_failed")
	newClient(t, ada.ID).get("/links/preview").expect(http.StatusUnprocessableEntity)
}

File: python/tests/linkpreview_test.py
"""Link preview / outbound HTTP (S-LP) acceptance scenarios, including the
catching halves of the G-HTTP case (timeout degradation, circuit breaker).
"""

from __future__ import annotations

import time

from harness.fixture import Fixture
from tests.helpers import client_for, seed_channel, seed_user


def test_s_lp_01_unfurl_success(fixture: Fixture) -> None:
    owner = seed_user(fixture, "lpowner")
    channel_id = seed_channel(fixture, owner, "general", private=False)
    fixture.unfurl.program_ok("Example")

    response = client_for(fixture, owner.id).post(
        f"/channels/{channel_id}/messages", {"text": "see https://example.com now"}
    )
    response.expect(201)
    assert response.json()["linkPreviewTitle"] == "Example"
    assert fixture.unfurl.request_count == 1


def test_s_lp_02_timeout_degrades_to_null(fixture: Fixture) -> None:
    """[G-HTTP] delay > 800 ms timeout → post 201 within 1.5 s, null preview."""
    owner = seed_user(fixture, "lpowner")
    channel_id = seed_channel(fixture, owner, "general", private=False)
    fixture.unfurl.program_delay(2.0)

    start = time.monotonic()
    response = client_for(fixture, owner.id).post(
        f"/channels/{channel_id}/messages", {"text": "slow https://example.com"}
    )
    elapsed = time.monotonic() - start
    response.expect(201)
    assert response.json()["linkPreviewTitle"] is None
    assert elapsed < 1.5, f"post took {elapsed:.2f}s — the timeout did not bound it"


def test_s_lp_03_server_error_degrades_to_null(fixture: Fixture) -> None:
    """[G-HTTP] upstream 500 → post 201, null preview."""
    owner = seed_user(fixture, "lpowner")
    channel_id = seed_channel(fixture, owner, "general", private=False)
    fixture.unfurl.program_server_error()
    response = client_for(fixture, owner.id).post(
        f"/channels/{channel_id}/messages", {"text": "broken https://example.com"}
    )
    response.expect(201)
    assert response.json()["linkPreviewTitle"] is None


def test_s_lp_04_circuit_breaker_opens(fixture: Fixture) -> None:
    """[G-HTTP] 5 failing posts (breaker opens) → 6th post 201 null preview AND
    stub request count == 5 (no 6th call)."""
    owner = seed_user(fixture, "lpowner")
    channel_id = seed_channel(fixture, owner, "general", private=False)
    fixture.unfurl.program_server_error()

    poster = client_for(fixture, owner.id)
    for i in range(6):
        response = poster.post(
            f"/channels/{channel_id}/messages", {"text": f"post {i} https://example.com"}
        )
        response.expect(201)
        assert response.json()["linkPreviewTitle"] is None
    assert fixture.unfurl.request_count == 5  # the breaker opened; the 6th never called out


def test_s_lp_05_direct_preview_proxy(fixture: Fixture) -> None:
    owner = seed_user(fixture, "lpowner")
    caller = client_for(fixture, owner.id)
    fixture.unfurl.program_ok("Proxied")
    ok = caller.get("/links/preview?url=https://example.com")
    ok.expect(200)
    assert ok.json()["title"] == "Proxied"

    fixture.unfurl.program_server_error()
    caller.get("/links/preview?url=https://example.com").expect(502).expect_code(
        "unfurl:upstream_failed"
    )
    caller.get("/links/preview?url=").expect(422)
